kitodo-mediaserver-core/src/main/java/org/kitodo/mediaserver/core/services/ActionService.java

```diff
@@ -0,0 +1,183 @@
+        }
+
+        actionData = new ActionData(work, actionName, parameter);
+        actionData.setRequestTime(Instant.now());
+        actionRepository.save(actionData);
+    }
+
+    /**
+     * Run a saved ActionData.
+     * @param work the Work
+     * @param actionName action bean name to be used
+     * @param parameter parameter list
+     * @return the result of the Action
+     * @throws Exception on Action errors
+     */
+    public Object performRequested(Work work, String actionName, Map<String, String> parameter) throws Exception {
+
+        // Does this action request exist?
+        ActionData actionData = getUnfinishedAction(work, actionName, parameter);
+        if (actionData == null) {
+            throw new ActionServiceException("This actionName='" + actionName + "' for workId='" + work.getId() + "' was not found.");
+        }
+
+        // Is this action already running?
+        ActionData persistentActionData = getRunningAction(work, actionName, parameter);
+        if (persistentActionData != null) {
+            throw new ActionServiceException("This actionName='" + actionName
+                + "' for workId=" + work.getId() + " is already running: startTime=" + persistentActionData.getStartTime());
+        }
+
+        Object actionInstance;
+
+        // Does the action bean exist?
+        try {
+            actionInstance = applicationContext.getBean(actionData.getActionName());
+        } catch (BeansException ex) {
+            throw new ClassNotFoundException("No Bean found for actionName='" + actionData.getActionName() + "'");
+        }
+
+        // Is the bean a usable Action?
+        if (!(actionInstance instanceof IAction)) {
+            throw new ActionServiceException("Saved actionName='" + actionData.getActionName()
+                + "' for workId=" + actionData.getWork().getId() + " is not a valid action bean.");
+        }
+
+        // Run the action...
+        actionData.setStartTime(Instant.now());
+        actionRepository.save(actionData);
+
+        Object result = ((IAction)actionInstance).perform(actionData.getWork(), actionData.getParameter());
+
+        actionData.setEndTime(Instant.now());
+        actionRepository.save(actionData);
+
+        return result;
+    }
+
+    /**
+     * Finds the last finished action.
+     * @param work the Work
+     * @param actionName the action bean name
+     * @return the ActionData
+     */
+    public ActionData getLastFinishedAction(Work work, String actionName) {
+        List<ActionData> actionDatas = actionRepository.findByWorkAndActionNameOrderByEndTimeDesc(work, actionName);
+        if (actionDatas != null && actionDatas.size() > 0) {
+            return actionDatas.get(0);
+        }
+        return null;
+    }
+
+    /**
+     * Finds the unfinished action (not running and running).
```

Comment:
The difference between unfinished and running action is a bit confusing, to me, unfinished suggests it is already running, and "not running and running" doesn't really explain it... Please write a few more words in the javadoc. (Suggestion: Finds a requested action that hasn't already been performed, whether it is currently running or not.) Maybe getUnperformedAction would be better...